File: behavioral/strategy.py
"""
Strategy (Policy)


Определяет семейство алгоритмов,
инкапсулирует каждый из них и делает их взаимозаменяемыми,
позволяет изменять алгоритмы независимо от клиентов, которые ими пользуются.
"""

from abc import ABCMeta, abstractmethod


class SalaryStrategy(metaclass=ABCMeta):
    """Стратегия оплаты труда."""

    @abstractmethod
    def calculate(self, hours_worked):
        """Считает размер заработной платы."""


class HourlyPaymentStrategy(SalaryStrategy):
    """Почасовая оплата труда."""

    def __init__(self, cost_per_hour):
        """
        Arguments:
            cost_per_hour (int): стоимость часа.
        """
        self._cost_per_hour = cost_per_hour

    def calculate(self, hours_worked):
        return round(self._cost_per_hour * hours_worked, 2)


class FixedRate(SalaryStrategy):
    """Фиксированная ставка."""

    DEFAULT_COUNT_WORKDAYS = 22

    def __init__(self, salary, work_day_duration=8):
        """
        Arguments:
            salary (float): размер оклада.
            work_day_duration (int): количество часов в рабочем дне.
        """
        self._salary = salary
        self._work_day_duration = work_day_duration

    def calculate(self, hours_worked):
        days = (hours_worked / self._work_day_duration)
        return round(self._salary * days / self.DEFAULT_COUNT_WORKDAYS, 2)


class Employee:
    """Сотрудник"""

    def __init__(self, firstname, lastname, hours_worked, salary_strategy):
        self._firstname = firstname
        self._lastname = lastname
        self._hours_worked = hours_worked
        self._salary_strategy = salary_strategy

    def get_fullname(self):
        """Возвращает полное имя сотрудника."""
        return f'{self._firstname} {self._lastname}'

    def calculate_salary(self):
        """Считает зарплату сотрудника."""
        return self._salary_strategy.calculate(self._hours_worked)


if __name__ == '__main__':
    employees = [
        Employee('Вася', 'Пупкин', 3, HourlyPaymentStrategy(1500)),
        Employee('Петя', 'Пупкин', 56, FixedRate(35000)),
    ]

    for person in employees:
        print(f'{person.get_fullname()} заработал {person.calculate_salary()}')

File: behavioral/strategy.php
<?php

/**
 * Strategy (Policy)
 *
 * Определяет семейство алгоритмов,
 * инкапсулирует каждый из них и делает их взаимозаменяемыми,
 * позволяет изменять алгоритмы независимо от клиентов, которые ими пользуются.
 */


/**
 * Стратегия оплаты труда
 */
interface SalaryStrategyInterface
{
    /**
     * Считает размер заработной платы.
     * 
     * @param int $hoursWorked количество отработанных часов.
     */
    public function calculate(int $hoursWorked): float;
}


/**
 * Почасовая оплата труда.
 */
class HourlyPaymentStrategy implements SalaryStrategyInterface
{
    private $costPerHour;

    /**
     * @param int $costPerHour стоимость часа.
     */
    public function __construct(int $costPerHour)
    {
        $this->costPerHour = $costPerHour;
    }

    public function calculate(int $hoursWorked): float
    {
        return round($this->costPerHour * $hoursWorked, 2);
    }
}
     

/**
 * Фиксированная ставка.
 */
class FixedRate implements SalaryStrategyInterface
{
    const DEFAULT_COUNT_WORKDAYS = 22;

    private $salary;
    private $workDayDuration;

    /**
     * @param float $salary размер оклада.
     * @param int $workDayDuration количество часов в рабочем дне.
     */
    public function __construct(float $salary, int $workDayDuration=8)
    {
        $this->salary = $salary;
        $this->workDayDuration = $workDayDuration;
    }

    public function calculate(int $hoursWorked): float
    {
        $days = $hoursWorked / $this->workDayDuration;
        return round($this->salary * $days / static::DEFAULT_COUNT_WORKDAYS, 2);
    }
}


/**
 * Сотрудник
 */
class Employee
{
    private $firstname;
    private $lastname;
    private $hoursWorked;
    private $salaryStrategy;

    public function __construct(
        string $firstname,
        string $lastname,
        int $hoursWorked,
        SalaryStrategyInterface $salaryStrategy
    ) {
        $this->firstname = $firstname;
        $this->lastname = $lastname;
        $this->hoursWorked = $hoursWorked;
        $this->salaryStrategy = $salaryStrategy;
    }

    /**
     * Возвращает полное имя сотрудника.
     */
    public function getFullname(): string
    {
        return $this->firstname . ' ' . $this->lastname;
    }

    /**
     * Считает зарплату сотрудника.
     */
    public function calculate_salary(): float
    {
        return $this->salaryStrategy->calculate($this->hoursWorked);
    }
}
        

$employees = [
    new Employee('Вася', 'Пупкин', 3, new HourlyPaymentStrategy(1500)),
    new Employee('Петя', 'Пупкин', 56, new FixedRate(35000)),
];

foreach ($employees as $person) {
    echo $person->getFullname() . ' заработал ' . $person->calculate_salary() . PHP_EOL;
}
